crates/composition/src/compose_supergraph.rs

```diff
@@ -32,14 +39,37 @@ fn merge_object_definitions<'a>(
         ));
     }
 
+    let first_is_entity = first.is_entity();
+    if definitions
+        .iter()
+        .any(|object| object.is_entity() != first_is_entity)
+    {
+        let name = first.name_str();
+        let (entity_subgraphs, non_entity_subgraphs) = definitions
+            .iter()
+            .partition::<Vec<DefinitionWalker<'_>>, _>(|definition| definition.is_entity());
+
+        ctx.diagnostics.push_fatal(format!(
+            "The `{name}` object is an entity in subgraphs {} but not in subgraphs {}.",
+            entity_subgraphs
+                .into_iter()
+                .map(|d| d.subgraph().name_str())
+                .join(", "),
+            non_entity_subgraphs
+                .into_iter()
+                .map(|d| d.subgraph().name_str())
+                .join(", "),
+        ));
+    }
```

Comment:
This is obviously not how we want the composition functions to look like in the end. I think I am starting to get a feel for how to organize them, but I'll keep implementing the basics before refactoring for the more complicated rules.